tests/python/rest_api/test_quality_control.py

```diff
@@ -619,36 +655,81 @@ def test_non_rq_job_owner_cannot_check_status_of_report_creation_in_sandbox(
                 u["id"] != task_staff["id"]
                 and not u["is_superuser"]
                 and u["id"] != task["owner"]["id"]
+                and u["id"] != (task["assignee"] or {}).get("id")
             )
         )
         rq_id = self._initialize_report_creation(task["id"], task_staff["username"])
-        self._test_check_status_of_report_creation_by_non_rq_job_owner(
+        self._test_check_status_of_report_creation(
             rq_id, task_staff=task_staff["username"], another_user=another_user["username"]
         )
 
-    @pytest.mark.parametrize("role", ("owner", "maintainer", "supervisor", "worker"))
-    def test_non_rq_job_owner_cannot_check_status_of_report_creation_in_org(
+    @pytest.mark.parametrize(
+        "same_org, role",
+        [
+            pair
+            for pair in product([True, False], _PermissionTestBase._default_org_roles)
+            if not (pair[0] and pair[1] in ["owner", "maintainer"])
+        ],
+    )
+    def test_user_without_rights_cannot_check_status_of_report_creation_in_org(
         self,
+        same_org: bool,
         role: str,
         admin_user: str,
         find_org_task_without_gt: Callable[[bool, str], tuple[dict[str, Any], dict[str, Any]]],
         find_users: Callable[..., list[dict[str, Any]]],
+        organizations,
     ):
         task, task_staff = find_org_task_without_gt(is_staff=True, user_org_role="supervisor")
 
         self.create_gt_job(admin_user, task["id"])
 
-        another_user = next(
-            u
-            for u in find_users(role=role, org=task["organization"])
-            if (
-                u["id"] != task_staff["id"]
-                and not u["is_superuser"]
-                and u["id"] != task["owner"]["id"]
+        org_filter = "org"
+        if not same_org:
+            org_filter = "exclude_" + org_filter
+
+        try:
+            another_user = next(
+                u
+                for u in find_users(
+                    role=role, exclude_is_superuser=True, **{org_filter: task["organization"]}
+                )
+                if (
+                    u["id"] != task_staff["id"]
+                    and u["id"] != task["owner"]["id"]
+                    and u["id"] != (task["assignee"] or {}).get("id")
+                )
             )
-        )
+        except StopIteration:
+            # create a new user that passes the requirements
```

Comment:
This logic is too complicated, IMO. You should either pre-create all necessary users via `data.json`, or always create the user in the test. Choosing what to do dynamically introduces unnecessary variance to the test.

Reply:
I prefer the second option since a task is defined dynamically based on several conditions. Using pre-created data may make this test unstable.